File: app/sdks/typescript/src/pagination.ts
/** Shape every /api/v1 list endpoint returns — see components.schemas.Pagination
 * in lib/api-v1/openapi.yaml and lib/api-v1/pagination.ts. */
export type Page<T> = {
  data: T[];
  pagination: { limit: number; nextCursor: string | null };
};

/**
 * Walks every page of a cursor-paginated list automatically, yielding one
 * item at a time — e.g. `for await (const audit of paginate((cursor) =>
 * client.audits.list({ cursor }))) { ... }`. The cursor is always passed
 * back exactly as received (opaque, per lib/api-v1/pagination.ts) — this
 * helper never inspects or constructs it.
 */
export async function* paginate<T>(fetchPage: (cursor?: string) => Promise<Page<T>>): AsyncGenerator<T, void, undefined> {
  let cursor: string | undefined;
  for (;;) {
    const page = await fetchPage(cursor);
    for (const item of page.data) yield item;
    if (!page.pagination.nextCursor) return;
    cursor = page.pagination.nextCursor;
  }
}

File: app/sdks/python/src/public_map_sdk/pagination.py
"""Cursor-pagination helper — see lib/api-v1/pagination.ts. The cursor is
always passed back exactly as received (opaque); this module never
inspects or constructs it."""

from __future__ import annotations

from typing import Callable, Generic, Iterator, Optional, TypeVar

from .models import Pagination

T = TypeVar("T")


class Page(Generic[T]):
    """Matches every /api/v1 list response's {data, pagination} shape."""

    def __init__(self, data: list, pagination: Pagination) -> None:
        self.data = data
        self.pagination = pagination


def paginate(fetch_page: Callable[[Optional[str]], "Page[T]"]) -> Iterator[T]:
    """Walks every page of a cursor-paginated list automatically, yielding
    one item at a time.

    >>> for audit in paginate(lambda cursor: client.audits.list(cursor=cursor)):
    ...     print(audit.id)
    """
    cursor: Optional[str] = None
    while True:
        page = fetch_page(cursor)
        for item in page.data:
            yield item
        if not page.pagination.next_cursor:
            return
        cursor = page.pagination.next_cursor
